plugins/inputs/logfile/logfile_test.go

```diff
@@ -358,100 +358,128 @@ func TestLogsFileRemove(t *testing.T) {
 	multilineWaitPeriod = 10 * time.Millisecond
-	logEntryString := "anything"
+
 	filePrefix := "file_auto_removal"
 	tmpfile1, err := createTempFile("", filePrefix)
+	fmt.Printf("Created 1st temp file, %s\n", tmpfile1.Name())
 	require.NoError(t, err)
 
-	_, err = tmpfile1.WriteString(logEntryString + "\n")
-	require.NoError(t, err)
-	tmpfile1.Sync()
-	tmpfile1.Close()
-
-	tt := NewLogFile()
-	tt.Log = TestLogger{t}
-	tt.FileConfig = []FileConfig{{
+	logFile := NewLogFile()
+	defer logFile.Stop()
+	logFile.Log = TestLogger{t}
+	logFile.FileConfig = []FileConfig{{
 		FilePath:      filepath.Join(filepath.Dir(tmpfile1.Name()), filePrefix+"*"),
 		FromBeginning: true,
 		AutoRemoval:   true,
 	}}
-	tt.FileConfig[0].init()
-	tt.started = true
+	logFile.FileConfig[0].init()
+	logFile.started = true
 
-	lsrcs := tt.FindLogSrc()
-	if len(lsrcs) != 1 {
-		t.Fatalf("%v log src was returned when 1 should be available", len(lsrcs))
-	}
+	logSource, evts := getLogSrc(t, logFile)
+	defer (*logSource).Stop()
 
-	lsrc := lsrcs[0]
-	evts := make(chan logs.LogEvent)
-	lsrc.SetOutput(func(e logs.LogEvent) {
-		if e != nil {
-			evts <- e
-		}
-	})
+	fmt.Println("Fill temp file with sufficient lines to be read.")
+	numLogLinesToWrite := 10000
+	logEntryString := "this is the best log line ever written to a file"
+	writeSomeLines(t, tmpfile1, numLogLinesToWrite, logEntryString)
+	tmpfile1.Close()
 
 	var tmpfile2 *os.File
+	var logSource2 *logs.LogSrc
+	var	evts2 chan logs.LogEvent
 	defer func() {
+		if logSource2 != nil {
+			(*logSource2).Stop()
+		}
 		if tmpfile2 != nil {
+			tmpfile2.Close()
 			os.Remove(tmpfile2.Name())
 		}
 	}()
 
-	wg.Add(1)
+	fmt.Println("Verify every line written to the first temp file is received.")
+	// Do this in a goroutine in case there is a bug and it hangs.
+	readerDone := make(chan bool)
+	// Need to make sure tmpfile2 is created before accessing evts2
+	fileCreatorDone := make(chan bool)
 
 	go func() {
-		defer wg.Done()
-
-		// create a new file matching configured pattern
-		tmpfile2, err = createTempFile("", filePrefix)
-		require.NoError(t, err)
-
-		_, err = tmpfile2.WriteString(logEntryString + "\n")
-		require.NoError(t, err)
-
+		defer close(readerDone)
+		for i := 0; i < numLogLinesToWrite; i++ {
+			logEvent := <- evts
+			require.Equal(t, logEntryString, logEvent.Message())
+			if i != numLogLinesToWrite / 2 {
+				continue
+			}
+			// Halfway through receiving events create a new temp file.
+			// Still expect to recv all events from the first file.
+			// Need to do this in a goroutine since FindLogSrc()
+			// will block until tailer reaches EOF on first temp file.
+			go func() {
+				defer close(fileCreatorDone)
+				tmpfile2, err = createTempFile("", filePrefix)
+				fmt.Printf("Created 2nd temp file, %s\n", tmpfile2.Name())
+				require.NoError(t, err)
+				_, err = tmpfile2.WriteString(logEntryString + "\n")
+				require.NoError(t, err)
+				logSource2, evts2 = getLogSrc(t, logFile)
+			}()
+		}
 	}()
```

Comment:
I'll look at this again in the morning, but this _looks_ like it could be pulled out/parameterized as a separate validation function - it would mess up some of the channel reads down below, but I don't think it would be terrible. At the end of the day, I'm not going to lose sleep if we have a unit test with 100 lines of code in it - skimming over this test, it is almost like an integration test in what it validates.

Reply:
I did some refactoring.
Split out code into some new functions.
`TestLogsFileAutoRemoval()` is now 14 lines long.
The longest helper function is 47 lines.